crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,742 @@
+                let start = end - TextSize::new(chars_to_remove);
+                diagnostic.set_fix(Fix::safe_edit(Edit::deletion(start, end)));
+
+                diagnostics.push(diagnostic);
+            }
+
+            if matches!(self.follows, Follows::Decorator) && line.preceding_blank_lines > 0 {
+                // E304
+                let mut diagnostic =
+                    Diagnostic::new(BlankLineAfterDecorator, line.first_token_range);
+
+                let range = line.first_token_range;
+                diagnostic.set_fix(Fix::safe_edit(Edit::deletion(
+                    locator.line_start(range.start())
+                        - TextSize::new(
+                            line.preceding_blank_characters
+                                .try_into()
+                                .expect("Number of blank characters to be small."),
+                        ),
+                    locator.line_start(range.start()),
+                )));
+
+                diagnostics.push(diagnostic);
+            }
+
+            if line.preceding_blank_lines < BLANK_LINES_TOP_LEVEL
+                && is_top_level_token(&self.previous_unindented_token)
+                && line.indent_level == 0
+                && !line.is_comment_only
+                && !is_top_level_token_or_decorator(&line.first_token)
+            {
+                // E305
+                let mut diagnostic = Diagnostic::new(
+                    BlankLinesAfterFunctionOrClass(line.blank_lines),
+                    line.first_token_range,
+                );
+
+                diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
+                    stylist
+                        .line_ending()
+                        .as_str()
+                        .to_string()
+                        .repeat((BLANK_LINES_TOP_LEVEL - line.blank_lines) as usize),
+                    locator.line_start(line.first_token_range.start()),
+                )));
+
+                diagnostics.push(diagnostic);
+            }
+
+            if line.preceding_blank_lines == 0
+            // Only apply to nested functions.
+                && matches!(self.fn_status, Status::Inside(_))
+                && is_top_level_token_or_decorator(&line.first_token)
+                // Allow following a decorator (if there is an error it will be triggered on the first decorator).
+                && !matches!(self.follows, Follows::Decorator)
+                // The class's docstring can directly precede the first function.
+                && !matches!(self.follows, Follows::Docstring)
+                // Do not trigger when the def/class follows an "indenting token" (if/while/etc...).
+                && prev_indent_level.is_some_and(|prev_indent_level| prev_indent_level >= line.indent_level)
+                // Allow groups of one-liners.
+                && !(matches!(self.follows, Follows::Def) && !matches!(line.last_token, Tok::Colon))
```

Comment:
Simply adding `indent_size` (which is always 4) won't produce the right result if a user uses e.g. 2 spaces indentation. 

I recommend changing `Inside` to store the indentation of the Class rather than the indentation of its body (removing the need to add `indent_size`). 

We know we're outside of the class if the `actual_indent <= class_indent`. 

Let's also add some tests with an indent other than 4 (2 spaces, tabs, mixed spaces and tabs)

Reply:
I've removed the hard coded indent size in 94c9babff.

I've added a few tests in 4f2c318a7, but they all passed from the start so that might not be enough.